compiler/lib/effects.ml

```diff
@@ -502,47 +579,41 @@ let cps_instr ~st (instr : instr) : instr =
   | Let (x, Apply { f; args; _ }) when not (Var.Set.mem x st.cps_needed) ->
       (* At the moment, we turn into CPS any function not called with
          the right number of parameter *)
-      assert (Global_flow.exact_call st.flow_info f (List.length args));
+      assert (
+        (* If this function is unknown to the global flow analysis, then it was
+           introduced by the lambda lifting and we don't have exactness info any more. *)
+        Var.idx f >= Var.Tbl.length st.flow_info.info_approximation
+        || Global_flow.exact_call st.flow_info f (List.length args));
       Let (x, Apply { f; args; exact = true })
-  | Let (_, (Apply _ | Prim (Extern ("%resume" | "%perform" | "%reperform"), _))) ->
+  | Let (_, e) when effect_primitive_or_application e ->
+      (* For the CPS target, applications of CPS functions and effect primitives require
+         more work (allocating a continuation and/or modifying end-of-block branches) and
+         are handled in a specialized function. *)
       assert false
   | _ -> instr
 
-let cps_block ~st ~k pc block =
+let cps_instr ~st (instr : instr) : instr list =
+  match instr with
+  | Let (x, Prim (Extern "caml_assume_no_perform", [ Pv f ])) when double_translate () ->
+      (* When double translation is enabled, we just call [f] in direct style.
+         Otherwise, the runtime primitive is used. *)
+      let unit = Var.fresh_n "unit" in
+      let exact = Global_flow.exact_call st.flow_info f 1 in
```

Comment:
We need to check everywhere that `Var.idx f < Var.Tbl.length st.flow_info.info_approximation`.
```suggestion
      let exact =
        Var.idx f < Var.Tbl.length st.flow_info.info_approximation
        && Global_flow.exact_call st.flow_info f 1
      in
```